jcl/src/openj9.sharedclasses/share/classes/com/ibm/oti/shared/SharedClassStatistics.java

```diff
@@ -45,69 +47,126 @@ public static long maxSizeBytes() {
-     * connected to.
-     * 
-     * @return the minimum shared classes cache space reserved for AOT data in bytes or -1 if it is not set.
-     */
-    public static long minAotBytes() {
-    	return minAotBytesImpl();
-    }
-    
-    /**
-     * Returns the maximum space allowed for AOT data of the cache that the JVM is currently
-     * connected to.
-     * 
-     * @return the maximum shared classes cache space allowed for AOT data in bytes or -1 if it is not set.
-     */
-    public static long maxAotBytes() {
-    	return maxAotBytesImpl();
-    }
-    
-    /**
-     * Returns the minimum space reserved for JIT data of the cache that the JVM is currently
-     * connected to.
-     * 
-     * @return the minimum shared classes cache space reserved for JIT data in bytes or -1 if it is not set.
-     */
-    public static long minJitDataBytes() {
-    	return minJitDataBytesImpl();
-    }
-    
-    /**
-     * Returns the maximum space allowed for JIT data of the cache that the JVM is currently
-     * connected to.
-     * 
-     * @return the maximum shared classes cache space allowed for JIT data in bytes  or -1 if it is not set.
-     */
-    public static long maxJitDataBytes() {
-    	return maxJitDataBytesImpl();
-    }
+	 * Returns the maximum space allowed for JIT data of the cache that the JVM is currently
+	 * connected to.
+	 *
+	 * @return the maximum shared classes cache space allowed for JIT data in bytes  or -1 if it is not set.
+	 */
+	public static long maxJitDataBytes() {
+		return maxJitDataBytesImpl();
+	}
 
+	/**
+	 * Returns the SysV shmem nattch value for a non-persistent top level cache.
+	 * On Windows or for persistent caches, -1 is returned. Depending on the platform
+	 * and OS, such as z/OS, the value indicates the number of JVMs attached to the cache.
+	 *
+	 * @return the SysV shmem nattch value for a top level non-persistent cache, or -1 if it cannot be determined.
+	 */
+	public static long numberAttached() {
+		return numberAttachedImpl();
+	}
+
+	/**
+	 * Returns the full cache path.
+	 *
+	 * @return the full cache path, or NULL if it cannot be determined.
+	 */
+	public static String cachePath() {
+		return cachePathImpl();
+	}
 	
-	private static native long maxSizeBytesImpl();
+	/**
+	 * Returns the cache name.
+	 *
+	 * @return the cache name, or NULL if it cannot be determined.
+	 */
+	public static String cacheName() {
+		String cachePath = cachePathImpl();
+		if (null != cachePath) {
+			File f = new File(cachePath);
+			String fileName = f.getName();
+			int prefixIndex = fileName.indexOf('_');
+			int tailIndex = fileName.lastIndexOf('_');
+			if ((prefixIndex > -1) && (tailIndex > -1)) {
+				String prefix = fileName.substring(0, prefixIndex);
+				String cacheName = fileName.substring(prefixIndex + 1, tailIndex);
+				if (!prefix.endsWith("P")) {
+					int typeIndex = cacheName.indexOf('_');
+					if (typeIndex > -1) {
+						cacheName = cacheName.substring(typeIndex + 1);
```

Comment:
What is the ultimate goal here? To extract the segment between the first and last underscores?
What is the significance of the prefix ending with 'P'?
Does anything break of the user says, for example, `-Xshareclasses:name=PHILIP`?

Reply:
The prefix ending with P means it's a persistent cache. When it's not a persistent cache the next section is `memory_`, which needs to be removed from the name. It doesn't matter what the actual name is (like `PHILIP`) as it's not part of the prefix.

Reply:
I've added a comment in the code.

Reply:
Example non-persistent cache name `C290M11F1A64_memory_test1_G43L00`

Reply:
If you add to the condition on line 132 `prefixIndex > 0`, then `prefix` is not needed and the test `!prefix.endsWith("P"` can be replaced by `fileName.charAt(prefixIndex - 1) != 'P'`.

Reply:
Updated as suggested.